### PR TITLE
Add Projector

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

## Summary

Add support for Projectors - an object that projects a texture onto a surface.

In order to get this new feature in gz-rendering7 in an ABI compatible way, I introduced a new `SceneExt` class that provides a generic `CreateExt` function for creating a `Projector` object.
  * The proper way is to add `CreateProjector` virtual function in base class and override them in ogre 1.x and 2.x implementation which changes vtable and so breaks ABI. I have added all these functions but they are commented out with a todo not to uncomment them when forward porting to `gz-rendering8`

Key implementation details:

* The ogre 1.x implementation is ported from gazebo-classic's [Projector](https://github.com/gazebosim/gazebo-classic/blob/gazebo11/gazebo/rendering/Projector.hh) class. It uses ogre material projective texturing feature.
     * in addition, I extended the implementation to support visibility flags  so that we can toggle the visibility of the projected texture based on the camera's visibility mask.
* The ogre 2.x implementation uses screen space decals (instead of projective texturing) and has a different behavior.
    * Also added support for setting visibility flags.     
    * **Note: There is one major caveat with this implementation.** The projection is not in the form of a "frustum" (i.e. projection becomes larger at longer distance). It uses ogre 2.x screen space decals. Think of a as a rectangular volume and any surface that intersects with the volume will have the texture mapped onto it. 
    * We will need to investigate whether it's possible to change the implementation to match the ogre 1.x behavior in the future. Here is a related [ogre forum post](https://forums.ogre3d.org/viewtopic.php?t=95298) asking the same question. I will create an issue to track this limitation once the PR is merged. 

The screenshot below illustrates the difference between ogre 1.x and ogre 2.x:

**ogre 1.x: Projection based on a frustum** (blue box is the projector)

projector_ogre

**ogre 2.x: Projection based on a rectangular volume**

projector_ogre2

## Test it

Run the `UNIT_Projector_TEST` and `INTEGRATION_projector` tests

e.g. testing with ogre 2.x

```sh
GZ_ENGINE_TO_TEST=ogre2 ./bin/UNIT_Projector_TEST
GZ_ENGINE_TO_TEST=ogre2 ./bin/INTEGRATION_projector
```

Build and run the `projector` demo

```sh
# build the demo
cd examples/projector
mkdir build
cd build
cmake ..
make

# Run the demo with ogre 1.x
./projector ogre

# Run the demo with ogre 2.x
./projector ogre2
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: examples/projector/GlutWindow.cc [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+
+#include <gz/common/Console.hh>
+#include <gz/math/Quaternion.hh>
+#include <gz/math/Vector3.hh>
+#include <gz/rendering/Camera.hh>
+#include <gz/rendering/Image.hh>
+#include <gz/rendering/Projector.hh>
+#include <gz/rendering/RayQuery.hh>
+#include <gz/rendering/Scene.hh>
+#include <gz/rendering/OrbitViewController.hh>
```
Comment (ahcorde) [inline examples/projector/GlutWindow.cc:42]: alphabetize

File: ogre2/src/Ogre2Projector.cc [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+{
+namespace rendering
+{
+inline namespace GZ_RENDERING_VERSION_NAMESPACE {
+//
+/// \brief Helper class for checking visibility of projecter to a camera
+class Ogre2ProjectorCameraListener: public Ogre::Camera::Listener
+{
+  /// \brief Constructor
+  /// \param[in] _node Pointer to the node that holds the decal (projector)
```
Comment (ahcorde) [inline ogre2/src/Ogre2Projector.cc:45]:
```suggestion
  /// \param[in] _decal Pointer to the node that holds the decal (projector)
```

File: ogre2/src/Ogre2Projector.cc [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+  if (this->VisibilityFlags() == GZ_VISIBILITY_ALL)
+  {
+    this->dataPtr->decalNode->setVisible(true);
+
+    for (auto &ogreCamIt : this->dataPtr->camerasWithListener)
+    {
+      Ogre::IdString camName = ogreCamIt.second;
+      // instead of getting the camera pointer through ogreCamIt.first,
+      // find camera pointer again to make sure the camera still exists
+      // because there is a chance that  we are holding onto a dangling pointer
```
Comment (ahcorde) [inline ogre2/src/Ogre2Projector.cc:174]:
```suggestion
      // because there is a chance that we are holding onto a dangling pointer
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: ogre2/src/Ogre2Projector.cc (hunk 2 above)
Comment (iche033) [reply ogre2/src/Ogre2Projector.cc]: [8383807](https://github.com/gazebosim/gz-rendering/pull/845/commits/8383807262ef97724ee90dbf0d51b887da8eb268)

File: examples/projector/Main.cc [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  #include <GL/glew.h>
+  #include <GL/gl.h>
+  #include <GL/glut.h>
+#endif
+
+#include <iostream>
+#include <vector>
+
+
+#include <gz/common/Console.hh>
```
Comment (ahcorde) [inline examples/projector/Main.cc:31]:
```suggestion
#include <vector>

#include <gz/common/Console.hh>
```

File: examples/projector/Main.cc [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        params["metal"] = "1";
+      }
+
+      CameraPtr camera = createCamera(engineName, params);
+      if (camera)
+        cameras.push_back(camera);
+    }
+    catch (...)
+    {
+      // std::cout << ex.what() << std::endl;
```
Comment (ahcorde) [inline examples/projector/Main.cc:213]:
```suggestion
```

File: include/gz/rendering/Scene.hh [hunk 6]
```diff
@@ -40,6 +42,8 @@ namespace gz
     inline namespace GZ_RENDERING_VERSION_NAMESPACE {
     //
     class RenderEngine;
+    class SceneExt;
+//    static std::unordered_map<const Scene *, SceneExt *> g_sceneExtMap;
```
Comment (ahcorde) [inline include/gz/rendering/Scene.hh:46]: remove?

**Review by ahcorde — APPROVED** [review]
Some minor comments, otherwise LGTM

File: src/base/BaseScene.cc [hunk 7]
```diff
@@ -1581,3 +1613,21 @@ void BaseScene::CreateMaterials()
   material->SetReceiveShadows(true);
   material->SetLightingEnabled(true);
 }
+
+/*
```
Comment (nkoenig) [inline src/base/BaseScene.cc:1617]: Did you want to add another comment about uncommenting this code in gz-rendering8?

File: ogre2/src/Ogre2Projector.cc [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+/////////////////////////////////////////////////
+Ogre2Projector::~Ogre2Projector()
+{
+  this->SetEnabled(false);
+
+  if (!this->scene->IsInitialized())
+    return;
+
+  for (auto &ogreCamIt : this->dataPtr->camerasWithListener)
```
Comment (nkoenig) [inline ogre2/src/Ogre2Projector.cc:118]: Should `ogreCamIt` be `const`?

**Review by nkoenig — APPROVED** [review]
Just minor comments

**Review by nkoenig — APPROVED** [review]
Just minor comments

File: ogre2/src/Ogre2Projector.cc (hunk 8 above)
Comment (iche033) [reply ogre2/src/Ogre2Projector.cc]: added `const`. [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

File: src/base/BaseScene.cc (hunk 7 above)
Comment (iche033) [reply src/base/BaseScene.cc]: oops, actually this code is no longer needed. Removed. [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

File: include/gz/rendering/Projector.hh [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      public virtual Visual
+    {
+      /// \brief Destructor
+      public: virtual ~Projector();
+
+      /// \brief Get the camera's far clipping plane distance
+      /// \return Far clipping plane distance
+      public: virtual double FarClipPlane() const = 0;
+
+      /// \brief Set the camera's far clipping plane distance
```
[the remark this answers is not in the document]
Comment (iche033) [reply include/gz/rendering/Projector.hh]: changed all instances of `camera` to `projector`. [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

File: include/gz/rendering/Projector.hh [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    /// gz/rendering/Projector.hh
+    //
+    /// \brief A projector that projects a texture onto a surface
+    class GZ_RENDERING_VISIBLE Projector :
+      public virtual Visual
+    {
+      /// \brief Destructor
+      public: virtual ~Projector();
+
+      /// \brief Get the camera's far clipping plane distance
```
[the remark this answers is not in the document]
Comment (iche033) [reply include/gz/rendering/Projector.hh]: fixed. [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

File: include/gz/rendering/Scene.hh (hunk 6 above)
Comment (iche033) [reply include/gz/rendering/Scene.hh]: removed, [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

File: examples/projector/Main.cc (hunk 5 above)
Comment (iche033) [reply examples/projector/Main.cc]: [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

File: examples/projector/Main.cc (hunk 4 above)
Comment (iche033) [reply examples/projector/Main.cc]: [b4a27fa](https://github.com/gazebosim/gz-rendering/pull/845/commits/b4a27fa08173162565dde561203ccc9273bf8126)

**Comment by ahcorde** [comment]
are all this windows failures expected ? https://build.osrfoundation.org/job/ign_rendering-pr-win/3223/

**Comment by iche033** [comment]
> are all these windows failures expected ? https://build.osrfoundation.org/job/ign_rendering-pr-win/3223/

yes unfortunately tests started failing recently on the new ec2 windows machines. I think @j-rivero is looking into it.

**Pull request merged**